src/metadata_encoding.rs

```diff
@@ -0,0 +1,203 @@
+use bytes::Bytes;
+use http::header::HeaderValue;
+use std::fmt;
+use std::error::Error;
+use std::hash::Hash;
+
+/// A possible error when converting a `MetadataValue` from a string or byte
+/// slice.
+#[derive(Debug, Default)]
+pub struct InvalidMetadataValue {
+    _priv: (),
+}
+
+pub trait ValueEncoding: Clone + Eq + PartialEq + Hash {
+    #[doc(hidden)]
+    fn is_valid_key(key: &str) -> bool;
+
+    #[doc(hidden)]
+    fn is_empty(value: &[u8]) -> bool;
+
+    #[doc(hidden)]
+    fn from_bytes(value: &[u8]) -> Result<HeaderValue, InvalidMetadataValueBytes>;
+
+    #[doc(hidden)]
+    fn from_shared(value: Bytes) -> Result<HeaderValue, InvalidMetadataValueBytes>;
+
+    #[doc(hidden)]
+    fn from_static(value: &'static str) -> HeaderValue;
+
+    #[doc(hidden)]
+    fn decode(value: &[u8]) -> Result<Bytes, InvalidMetadataValueBytes>;
+
+    #[doc(hidden)]
+    fn equals(a: &HeaderValue, b: &[u8]) -> bool;
+
+    #[doc(hidden)]
+    fn values_equal(a: &HeaderValue, b: &HeaderValue) -> bool;
+
+    #[doc(hidden)]
+    fn fmt(value: &HeaderValue, f: &mut fmt::Formatter) -> fmt::Result;
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Ascii {}
```

Comment:
If these are just marker types, and should never be instantiated, consider making them empty enums.

Reply:
Nice idea, will do.